### PR TITLE
SP: Attempt to fix GetString() calls with cvars in ICARUS

### PR DESCRIPTION
**Pull request by ensiform**
Cvar_VariableStringBuffer() does not like unallocated buffers with strlen().

Uses a local variable and cache the result in a map<> that is then cleared upon shutdown.

### CODE REVIEW

**Comment by mrwonko** [comment]
Adding a cache feels weird. Caches usually make things more complicated. I'm not familiar enough with the code to tell if they're necessary at a quick glance, I need to study the code in detail first.

How do I reproduce the issue this fixes?

**Comment by AntiAnti** [comment]
> Adding a cache feels weird. Caches usually make things more complicated. I'm not familiar enough with the code to tell if they're necessary at a quick glance, I need to study the code in detail first.
> 
> How do I reproduce the issue this fixes?

Read any cvar in script. For example:
`print ( $get( STRING, "cvar_g_char_model")$ );`

There is another possible, simplier fix for this bug, but it wouldn't work for two cvar requests in the same expression (for example, in the IF expression). Not critical, as for me.
```
// Q3_Interface.cpp, CQuake3GameInterface::GetString, line 10571
if( strlen(name) > 5 &&
	Q_stricmpn(name, "cvar_", 5) )
{
	static char cvarbuf[MAX_STRING_CHARS];

	// save in static buffer rather then NULL pointer
	gi.Cvar_VariableStringBuffer(name+5, cvarbuf, sizeof(cvarbuf));
	*value = cvarbuf;
	return true;
}
```

I need to fix this bug for my mod, which is almost done now.

**Review by mrwonko — APPROVED** [review]
This should fix the issue it tries to, and doesn't really do anything wrong, even if it feels a bit weird to copy the value.

**Comment by ensiform** [comment]
Are you saying the cache code still crashes?

The point of the cache code is to not reuse one buffer and clobber over it in cases of multiple calls.

**Comment by mrwonko** [comment]
> Are you saying the cache code still crashes?

No, just that caches are naturally complex and best avoided. But given the constraints, it's acceptable here, it doesn't prevent re-evaluation, it's just a form of memory management.

File: code/game/Q3_Interface.cpp [hunk 1]
```diff
@@ -10568,10 +10568,13 @@ int		CQuake3GameInterface::GetString( int entID, const char *name, char **value
...
-	if( !Q_stricmpn(name, "cvar_", 5) &&
-		strlen(name) > 5 )
+	if( strlen(name) > 5 && Q_stricmpn(name, "cvar_", 5) )
 	{
-		gi.Cvar_VariableStringBuffer(name+5, *value, strlen(*value));
+		char cvarbuf[MAX_STRING_CHARS];
+
+		gi.Cvar_VariableStringBuffer(name+5, cvarbuf, sizeof(cvarbuf));
+		m_cvars[name+5] = cvarbuf;
+		*value = &m_cvars[name+5][0];
```
[the remark this answers is not in the document]
Comment (ensiform) [reply code/game/Q3_Interface.cpp:10577]: That very well may be, but as I don't have scripts that can vigorously test the corner cases and such hard to say. and I don't really feel like putting in the time to do so either lol

[the remark this answers is not in the document]
Comment (mrwonko) [reply code/game/Q3_Interface.cpp:10577]: I'll give it another look tonight.

**Review by mrwonko — CHANGES_REQUESTED** [review]
The potential use-after-free scares me, I'll take a stab at fixing it later

File: code/game/Q3_Interface.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (mrwonko) [reply code/game/Q3_Interface.cpp]: I've fixed it by replacing the `std::string` map values with fixed-size buffers, which then get re-used on subsequent calls, so the pointers never become invalid.

File: code/game/Q3_Interface.cpp [hunk 2]
```diff
@@ -26,6 +26,7 @@ along with this program; if not, see <http://www.gnu.org/licenses/>.
 //	is not directly portable from engine to engine.
 //
 //	-- jweier
+#include <array>
```
Comment (mrwonko) [inline code/game/Q3_Interface.cpp:29]: I'm not entirely sure about the right place to include <array> according to q3 conventions, so I used normal c++ best practices and only included it in the .cpp that actually needs it

**Review by mrwonko — APPROVED** [review]

**Comment by mrwonko** [comment]
I've formally approved this, but since I also contributed code changes I'd prefer a second opinion before we merge

**Review by bartrpe — APPROVED** [review]

**Comment by ensiform** [comment]
Go ahead
